pyro/distributions/lkj.py

```diff
@@ -0,0 +1,209 @@
+
+
+corr_cholesky_constraint = _CorrCholesky()
+
+
+########################################
+# Define transforms
+########################################
+
+def _vector_to_l_cholesky(z):
+    D = (1.0 + math.sqrt(1.0 + 8.0 * z.shape[-1]))/2.0
+    if D % 1 != 0:
+        raise ValueError("Correlation matrix transformation requires d choose 2 inputs")
+    D = int(D)
+    x = z.new_zeros(list(z.shape[:-1]) + [D, D])
+
+    x[..., 0, 0] = 1
+    x[..., 1:, 0] = z[..., :(D-1)]
+    i = D - 1
+    last_squared_x = z.new_zeros(list(z.shape[:-1]) + [D])
+    for j in range(1, D):
+        distance_to_copy = D - 1 - j
+        last_squared_x = last_squared_x[..., 1:] + x[..., j:, (j-1)].clone()**2
+        x[..., j, j] = (1 - last_squared_x[..., 0]).sqrt()
+        x[..., (j+1):, j] = z[..., i:(i + distance_to_copy)] * (1 - last_squared_x[..., 1:]).sqrt()
+        i += distance_to_copy
+    return x
+
+
+class CorrLCholeskyTransform(Transform):
+    """
+    Transforms a vector into the cholesky factor of a correlation matrix.
+
+    The input should have shape `[batch_shape] + [d * (d-1)/2]`. The output will have
+    shape `[batch_shape] + [d, d]`.
+
+    Reference:
+
+    [1] `Cholesky Factors of Correlation Matrices`, Stan Reference Manual v2.18, Section 10.12
+    """
+    domain = constraints.real
+    codomain = corr_cholesky_constraint
+    bijective = True
+    sign = +1
+    event_dim = 1
+
+    def __eq__(self, other):
+        return isinstance(other, CorrLCholeskyTransform)
+
+    def _call(self, x):
+        z = x.tanh()
+        return _vector_to_l_cholesky(z)
+
+    def _inverse(self, y):
+        if (y.shape[-2] != y.shape[-1]):
+            raise ValueError("A matrix that isn't square can't be a Cholesky factor of a correlation matrix")
+        D = y.shape[-1]
+
+        z_tri = y.new_zeros(y.shape[:-2] + (D - 2, D - 2))
+        z_stack = [
+            y[..., 1:, 0]
+        ]
+
+        for i in range(2, D):
+            z_tri[..., i - 2, 0:(i-1)] = y[..., i, 1:i] / (1-y[..., i, 0:(i-1)].pow(2).cumsum(-1)).sqrt()
+        for j in range(D - 2):
+            z_stack.append(z_tri[..., j:, j])
+
+        z = torch.cat(z_stack, -1)
+        return torch.log1p((2*z)/(1-z)) / 2
+
+    def log_abs_det_jacobian(self, x, y):
+        # Note dependence on pytorch 1.0.1 for batched tril
+        tanpart = x.cosh().log().sum(-1).mul(-2)
+        matpart = (1 - y.pow(2).cumsum(-1).tril(diagonal=-2)).log().div(2).sum(-1).sum(-1)
+        return tanpart + matpart
+
+# register transform to global store
+@biject_to.register(corr_cholesky_constraint)
+@transform_to.register(corr_cholesky_constraint)
+def _transform_to_corr_cholesky(constraint):
+    return CorrLCholeskyTransform()
+
+
+########################################
+# Define distribution
+########################################
+
+# TODO: Modify class to support more than one eta value at a time?
+class CorrLCholeskyLKJPrior(TorchDistribution):
```

Comment:
On a completely unrelated note, what do you think about renaming this to `LKJCorrPrior` or anything short enough that begins with `LKJ`? A simple name would be memorable and less wordy to use in day to day. That this implements a cholesky version can be read up in the docstrings, and is standard enough in other libraries that we probably can skip it in the name.